Add Transformation from Separator to Spacer

rudrakshi-gupta opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
This PR introduces a transform for converting `Separator` blocks into `Spacer` blocks, providing better control over white space in content.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
This PR addresses the issue of outdated content that contains excessive use of hr elements, which are being converted into `Separator` blocks during migration to the Block Editor. Many users relied on hr for spacing, but transitioning to `Spacer` blocks gives more control and helps achieve a cleaner, modern design with better-managed whitespace.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
A transform has been added to convert `Separator` blocks into `Spacer` blocks. Specifically, the transform generates a Spacer block with a default height of 50px, replacing the Separator block and offering more flexibility in controlling spacing between content sections

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
1. **Open the Block Editor**: Start by opening a post or page in the Block Editor.
2. **Insert a `Separator` Block**: 
   - Use the block inserter (the + icon) to search for and add a `Separator` block to your content.
3. **Transform the Block**: 
   - With the `Separator` block selected, look at the block toolbar (just above the block).
   - Click on the "Transform" button (usually represented by two arrows or the option to change block type).
   - From the dropdown, select `Spacer` to transform the `Separator` block into a `Spacer` block.
4. **Verify the Transformation**: 
   - Check that the `Spacer` block is correctly added and that the default height is set to 50px. You can view this in the block settings on the right-hand panel.

Feel free to explore further by adjusting the height in the block inspector to ensure everything works as expected!

## Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
1. **Open the Block Editor**: Navigate to a post or page using the **Tab** key to reach the post/page listing, then press **Enter** to open the editor.
2. **Insert a `Separator` Block**: 
   - Press **Tab** to focus on the block inserter button (+ icon).
   - Press **Enter** to open the block inserter.
   - Use the arrow keys to navigate to the `Separator` block.
   - Press **Enter** to insert the `Separator` block.
3. **Focus on Block Controls**: 
   - Press **Tab** to navigate to the block toolbar (above the block).
   - Use the **Tab** key to focus on the "Transform" button (it usually appears as a block change option).
   - Press **Enter** to open the transform options.
4. **Transform to Spacer Block**: 
   - Use the arrow keys to navigate to the `Spacer` block option.
   - Press **Enter** to transform the `Separator` block into a `Spacer` block.
5. **Check Spacer Properties**: 
   - Press **Tab** to focus on the block inspector panel (right sidebar).
   - Use the arrow keys to navigate through the `Spacer` block settings and verify that the height has been set to `50px`.

## Screenshots or screencast 
<img width="642" alt="Screenshot 2024-10-18 at 12 53 34 PM" src="https://github.com/user-attachments/assets/6e09690d-d921-4efc-8c5e-41a8cbc739a9">

Commit:
Add a transform from the Separator block to the Spacer block

Commit:
Merge branch 'trunk' into feat/seprator-transform-spacer
Merge latest updates of trunk to feat/seprator-transform-spacer branch

Commit:
Add a transform from the Separator block to the Spacer block

Feat #65492

talldan commented on packages/block-library/src/separator/transforms.js:
```diff
@@ -18,6 +18,17 @@ const transforms = {
+			blocks: [ 'core/spacer' ], // Transform to Spacer.
+			transform: () => {
+				return createBlock( 'core/spacer', {
+					height: '50px',
```
Is the `50px` to achieve a height similar to a separator? The spacer does already have a `100px` default.

It'd be good to include any other mutual attributes. I think `anchor` is the only on that I can see. If a separator has an anchor it should be retained when transforming to the spacer.

rudrakshi-gupta replied:
Thank you for the feedback!

1. The 50px height was an arbitrary choice to mimic a slightly smaller spacing effect than the Spacer block's default 100px. However, I understand that 100px is the default height for a Spacer, so I'm happy to adjust it to the default if that better fits the expected behavior.
2. Great point about the anchor attribute! I will update the transform so that if a Separator block has an anchor, it will be retained when transforming to the Spacer block. This ensures that the transformation keeps user-defined anchors intact.

talldan reviewed (commented):
Thanks for your contribution and the very well written PR description.

I think the one thing missing that would be good to add is a transform the other way from spacer to separator. Usually as a rule of thumb blocks can be transformed both ways.

rudrakshi-gupta commented:
> Thanks for your contribution and the very well written PR description.
> 
> I think the one thing missing that would be good to add is a transform the other way from spacer to separator. Usually as a rule of thumb blocks can be transformed both ways.

Thanks for the feedback! I agree, it makes sense to include a transformation from spacer to separator as well. I'll work on adding that functionality to ensure the blocks can be transformed both ways.

Commit:
Add transformation from spacer to separator

- Remove the default height of 50px after transformation to default 100px that of Spacer
- Update the transform of blocks so that it retains anchor attribute

talldan commented:
Thanks for making those changes. This is looking good to me, except it looks like there are unit tests failing (https://github.com/WordPress/gutenberg/actions/runs/11434042367/job/31806942028?pr=66230).

It should hopefully only be a matter of adding the block titles of the new transform options to these places and then this PR will be good to merge:
- https://github.com/WordPress/gutenberg/blob/2fe909bbb60a07cb4d9c026d5d63f2ea8098e1c0/packages/block-library/src/spacer/test/transforms.native.js#L18
- https://github.com/WordPress/gutenberg/blob/2fe909bbb60a07cb4d9c026d5d63f2ea8098e1c0/packages/block-library/src/separator/test/transforms.native.js#L18

Commit:
Add new transformation block titles in Test

Commit:
Add new transformation block titles in Test

Commit:
Add new transformation block titles in Test and resolve linting problem

Commit:
Add new transformation block titles in Test and resolve linting problem

Commit:
Update test suite snapshot for test [mobile] to pass successfully

rudrakshi-gupta commented:
@talldan Thanks for the feedback! 

I’ve added the new block titles to both `transforms.native.js` files for the Spacer and Separator blocks, including the necessary updates to the `transformsWithInnerBlocks` arrays. All tests are now passing after the updates.

Please let me know if there’s anything else needed!

talldan reviewed (approved):
Thanks for all your work here!

(Congrats on your first PR contribution!)
